src/server/game/Entities/AreaTrigger/AreaTrigger.cpp

```diff
@@ -96,13 +157,386 @@ void AreaTrigger::Update(uint32 p_time)
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    float halfExtentsX = extentsX / 2.0f;
+    float halfExtentsY = extentsY / 2.0f;
+    float halfExtentsZ = extentsZ / 2.0f;
+
+    float minX = GetPositionX() - halfExtentsX;
+    float maxX = GetPositionX() + halfExtentsX;
+
+    float minY = GetPositionY() - halfExtentsY;
+    float maxY = GetPositionY() + halfExtentsY;
+
+    float minZ = GetPositionZ() - halfExtentsZ;
+    float maxZ = GetPositionZ() + halfExtentsZ;
+
+    G3D::AABox box({ minX, minY, minZ }, { maxX, maxY, maxZ });
+
+    targetList.remove_if([this, box](Unit* unit) -> bool
+    {
+        return !box.contains({ unit->GetPositionX(), unit->GetPositionY(), unit->GetPositionZ() });
+    });
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::SearchUnitInPolygon()
+{
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius, false);
+    Trinity::UnitListSearcher<Trinity::AnyUnitInObjectRangeCheck> searcher(this, targetList, check);
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    float height = GetTemplate()->PolygonDatas.Height;
+
+    UpdatePolygonOrientation();
+
+    targetList.remove_if([this, height](Unit* unit) -> bool
+    {
+        float minZ = GetPositionZ() - height;
+        float maxZ = GetPositionZ() + height;
+
+        return  !CheckIsInPolygon2D(unit) ||
+                unit->GetPositionZ() < minZ ||
+                unit->GetPositionZ() > maxZ;
+    });
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::SearchUnitInCylinder()
+{
+    float height = GetTemplate()->CylinderDatas.Height;
+
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius, false);
+    Trinity::UnitListSearcher<Trinity::AnyUnitInObjectRangeCheck> searcher(this, targetList, check);
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    targetList.remove_if([this, height](Unit* unit) -> bool
+    {
+        float minZ = GetPositionZ() - height;
+        float maxZ = GetPositionZ() + height;
+
+        return unit->GetPositionZ() < minZ || unit->GetPositionZ() > maxZ;
+    });
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::HandleUnitEnterExit(std::list<Unit*> newTargetList)
+{
+    GuidUnorderedSet exitUnits = _insideUnits;
+    _insideUnits.clear();
+
+    for (Unit* unit : newTargetList)
+    {
+        if (exitUnits.erase(unit->GetGUID()) == 0) // erase(key_type) returns number of elements erased
+        {
+            if (Player* player = unit->ToPlayer())
+                if (player->isDebugAreaTriggers)
+                    ChatHandler(player->GetSession()).PSendSysMessage("You entered areatrigger %u", GetTemplate()->Id);
+
+            AddAuras(unit);
+            sScriptMgr->OnAreaTriggerEntityUnitEnter(this, unit);
```

Comment:
you can't use GetInsideUnits() in this hook, i think it is better so add the unit to a enterUnits container and handle hooks and auras after all units are reinserted in _insideUnits